homeassistant/components/firmata/pin.py

```
@@ -0,0 +1,145 @@
+        super().__init__(board, pin, pin_mode)
+
+    async def start_pin(self, forward_callback):
+        """Get initial state and start reporting a pin."""
+        _LOGGER.debug(
+            "Starting reporting updates for input pin %s on board %s",
+            self._pin,
+            self._board.name,
+        )
+        self._forward_callback = forward_callback
+        api = self._board.api
+        if self._pin_mode == PIN_MODE_INPUT:
+            await api.set_pin_mode_digital_input(self._pin, self.latch_callback)
+        elif self._pin_mode == PIN_MODE_PULLUP:
+            await api.set_pin_mode_digital_input_pullup(self._pin, self.latch_callback)
+
+        new_state = bool((await self._board.api.digital_read(self._firmata_pin))[0])
+        if self._negate:
+            new_state = not new_state
+        self._state = new_state
+
+        await api.enable_digital_reporting(self._pin)
+        self._forward_callback()
+
+    async def stop_pin(self):
+        """Stop reporting digital input pin."""
+        _LOGGER.debug(
+            "Stopping reporting updates for digital input pin %s on board %s",
+            self._pin,
+            self._board.name,
+        )
+        api = self._board.api
+        await api.disable_digital_reporting(self._pin)
+
+    @property
+    def is_on(self) -> bool:
+        """Return true if digital input is on."""
+        return self._state
+
+    @callback
+    async def latch_callback(self, data):
+        """Update pin state on callback."""
+        if data[1] != self._firmata_pin:
+            return
+        _LOGGER.debug(
+            "Received latch %d for digital input pin %d on board %s",
+            data[2],
+            self._firmata_pin,
+            self._board.name,
+        )
+        new_state = bool(data[2])
+        if self._negate:
+            new_state = not new_state
+        if self._state == new_state:
+            return
+        self._state = new_state
+        self._forward_callback()
```

Should we raise a specific error if the callback isn't set?

---

Sure, I can add it, though that error would only be called as a result of a very direct programming error that is isolated to this component. This is only affected by calling `start_pin()` without the callback argument, which is only called by the entity. `latch_callback()` is never called before `start_pin()` is called.

---

Ok, let's keep it like this and just blow up. That's fine.